fedimint-api/src/config.rs

```diff
@@ -19,41 +21,187 @@ use tbs::hash::hash_bytes_to_curve;
+/// and better UX, while the non-standard modules have to use a type-erased
+/// config.
+///
+/// Candidate for re-designing when the modularization effort is
+/// complete.
+pub struct ModuleConfigGenParams {
+    pub mint_amounts: Vec<Amount>,
+    pub bitcoin_rpc: BitcoindRpcCfg,
+
+    /// extra options for extra settings and modules
+    pub other: BTreeMap<String, serde_json::Value>,
+}
+
+impl ModuleConfigGenParams {
+    /// Default & fake config gen params for things like tests
+    ///
+    /// TODO: Possibly this does not belong here.
+    pub fn fake_config_gen_params() -> ModuleConfigGenParams {
+        ModuleConfigGenParams {
+            mint_amounts: [1, 10, 100, 1000, 10000, 100000, 1000000]
+                .into_iter()
+                .map(Amount::from_milli_sats)
+                .collect(),
+            bitcoin_rpc: fedimint_api::config::BitcoindRpcCfg {
+                btc_rpc_address: "localhost".into(),
+                btc_rpc_user: "bitcoin".into(),
+                btc_rpc_pass: "bitcoin".into(),
+            },
+            other: Default::default(),
+        }
+    }
+}
+
+/// Config for the client-side of a particular Federation module
+///
+/// Since modules are (tbd.) pluggable into Federations,
+/// it needs to be some form of an abstract type-erased-like
+/// value.
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct ClientModuleConfig(pub serde_json::Value);
+
+impl From<serde_json::Value> for ClientModuleConfig {
+    fn from(v: serde_json::Value) -> Self {
+        Self(v)
+    }
+}
+
+impl ClientModuleConfig {
+    pub fn cast<T: TypedClientModuleConfig>(&self) -> anyhow::Result<T> {
+        Ok(serde_json::from_value(self.0.clone())?)
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct ServerModuleConfig(pub serde_json::Value);
+
+impl From<serde_json::Value> for ServerModuleConfig {
+    fn from(v: serde_json::Value) -> Self {
+        Self(v)
+    }
+}
+
+impl ServerModuleConfig {
+    pub fn cast<T: TypedServerModuleConfig>(&self) -> anyhow::Result<T> {
+        Ok(serde_json::from_value(self.0.clone())?)
+    }
+}
+
+pub trait TypedServerModuleConfig: DeserializeOwned + Serialize {
+    fn to_erased(&self) -> ServerModuleConfig {
+        ServerModuleConfig(serde_json::to_value(&self).expect("serialization can't fail"))
+    }
+
+    fn to_client_config(&self) -> ClientModuleConfig;
+
+    fn validate_config(&self, identity: &PeerId) -> anyhow::Result<()>;
+}
+
+pub trait TypedClientModuleConfig: DeserializeOwned + Serialize {
+    fn to_erased(&self) -> ClientModuleConfig {
+        ClientModuleConfig(serde_json::to_value(&self).expect("serialization can't fail"))
+    }
+}
+
+/// Things that a `distributed_gen` config can send between peers
+#[derive(Serialize, Deserialize, Debug, Clone)]
+// TODO: is this even needed?
+#[non_exhaustive]
+pub enum ConfigGenPeerMsg {
+    PublicKey(secp256k1::PublicKey),
```

Comment:
Rename after how it is used.

Reply:
I unfortunately don't know enough about this off the top of my head, maybe @jkitman can help

Reply:
`PublicKey` is probably the pubkey for the bitcoin wallet module so maybe it could be `WalletPublicKey`

`DistributedGen` is any message for DKG so maybe `DkgMessage` is appropriate.

The `(String, SupportedDpkgMsg)` can be refactored away once we can use `AnyPeerConnections` for multiple DKGs.

Reply:
The String here is for mint amount tiers, and is unused for other types of messages. :D